### ISSUE TITLE
Can't connect to synapse with weechat-matrix, with mirage it works.

### ISSUE THREAD

**Issue by hashashini** [poljar/weechat-matrix#192]
When trying to connect to my homeserver I get the following error:
```
nio.client: Error with response of type type: <class 'nio.responses.LoginInfoResponse'>, error code 404
nio.responses: Validating response schema
nio.responses: Error validating response: 'flows' is a required property
nio.client: Received new response of type LoginInfoError
matrix: Error: LoginInfoError: 404 unknown error
```
It looks like #168, and indeed,
`curl https://$my_homeserver/_matrix/login`
returns
```
<html>
  <head><title>404 - No Such Resource</title></head>
  <body>
    <h1>No Such Resource</h1>
    <p>Sorry. No luck finding that resource.</p>
  </body>
</html>
```

But
`curl https://$my_homeserver/_matrix/client/r0/login`
works:
```
{
    "flows": [
        {
            "type": "m.login.password"
        }
    ]
}
```
And as far as I understand https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-login that's actually the expected behaviour conforming to the specification. Am I missing something?

Regards!

**Comment by poljar** [comment]
That comment is wrong, it [does](https://github.com/poljar/matrix-nio/blob/master/nio/api.py#L243) use the `/client/r0/login` path.

**Comment by hashashini** [comment]
Sorry, didn't mean to offend and apparently shouldn't have drawn such a conclusion, especially since I'm new to matrix.

Does this mean that the problem is with my setup? If so, what can I do that synapse answers the `curl https://$my_homeserver/_matrix/login` request correctly?

With Mirage I'm able to successfully connect to my synapse homeserver behind a reverse proxy.
Maybe this line is a hint, too; it appears shortly before the error messages:
`nio.http: Got http header server: Werkzeug/0.16.1 Python/3.8.2`
Werkzeug is a server that runs on a different machine behind the reverse proxy.
But 'curl https://$my_homeserver/_matrix/client/r0/login' seems to work as expected, doesn't it? With riot.im and riotx I don't have problems connecting either.

I'd appreciate any help.

**Comment by poljar** [comment]
No offense taken :D. Weechat-matrix is pretty dumb with regards to discovering the right server. It really doesn't do any discovery that it should,  you'll need to point it at the right hostname/port. Using `curl -v ...` might reveal where that is.

**Comment by hashashini** [comment]
It's subdomain.domain.tld:443
`curl -v subdomain.domain.tld` returns
```
< Location: /_matrix/static
< Server: Synapse/1.12.2
```
That's my configuration in weechat-matrix:
```
address. . . : subdomain.domain.tld
port . . . . :   (443)
```

Does weechat-matrix in this case send a header "Host" with value "subdomain.domain.tld"?

I also already tried what would involve discovery via .well-known/matrix/client, to no avail:
```
address. . . : domain.tld
port . . . . :   (443)
```

**Comment by poljar** [comment]
I think you'll have to use the `subdomain.domain.tld` one but append the correct path as well, if I understood what `/_matrix/static` is.

You might want to connect weechat-matrix trough [mitmproxy](https://mitmproxy.org/) to check out further where it's going wrong.

**Comment by hashashini** [comment]
When I visit `subdomain.domain.tld` in a browser, `subdomain.domain.tld/_matrix/static` displays synapse's "It works! Synapse is running" landing page.
Nevertheless I tried `/set matrix.server.domain.address subdomain.domain.tld/_matrix/static`, but had no success.

My reverse proxy forwards traffic to synapse only when it encounters a `Host` header with value `subdomain.domain.tld` as part of the request. Requests to `domain.tld` are forwarded to a different server. I already tried sniffing the traffic with wireshark, but the plain text part didn't reveal anything helpful. I'll continue to investigate the issue tomorrow. Thank you so far!
